Remove Serving Node from architecture

michelav opened the pull request:
Removing Serving Node from architecture Documentation. Also moved Lab1 to README.

It fixes #25

michelav committed:
Remove Serving Node from architecture design

michelav committed:
Move Lab file to README

michelav committed:
Remove Serving Node mentions from architecture description

michelav committed:
Sort team members alphabetically

anayden commented on Lab1_Operationalizing_Pytorch_with_Mlflow/README.md:
```diff
@@ -1,31 +1,127 @@
-- Team 3 : Laszlo Sragner, yartem, Dimi, Paulo Maia
-- Team 4 : Shadab Hussain, susana brandao, Godwin Ekainu, Josh
+Following list describes the most relevant folders in which this lab is organized. Some folders in 
+this repo are self-explanatory (like [_imgs_](imgs/)) and therefore aren'r mentioned in the list.
```
super minor: typo, `aren'r`

michelav replied:
Done!

anayden commented on Lab1_Operationalizing_Pytorch_with_Mlflow/README.md:
```diff
@@ -1,31 +1,127 @@
+_MLFlow Tracking Server_ hosted in the **Tracking Node**. All these items, metadata and trained models, 
+are stored in external resources: RDBMS and Cloud Storage Services. Trained models are embedded into 
+dockerized **Serving Instances**, containers running [Torchserve](https://pytorch.org/serve/) providing 
+web access to the models.
```
minor: I would say `providing access to the models via HTTP`, because there's no web UI per se.

michelav replied:
Yes... pretty much better! ;-)

anayden commented on Lab1_Operationalizing_Pytorch_with_Mlflow/README.md:
```diff
@@ -1,31 +1,127 @@
+web access to the models.
+
+In order to provide feedback loop, a **Monitor Node** checks the status and performance meters, issuing 
+new training cycles or a full rebuild process. Some structures traverses through Development and Production perimeters. Training and Monitor nodes are key elements that communicate between those zones. 
```
> Some structures traverses through Development and Production perimeters

I don't get this phrase. Would you care to elaborate?

michelav replied:
Usually, when we think about "system architecture" we try to isolate elements in order to comply with security requirements. Something like Development traffic never reaches Production zone, that is, Development nodes would "lie" in Development zone with no connectivity with nodes \ servers in Production. However, when I drew that, I was thinking a "peculiar" Monitoring Node. One that could reach Development and Production elements. It would lie in between zones. Just like a "cross-cutting concern". 

I was trying to (concisely) say that with that phrase. However, my summarization skills are flawed somehow. hahahahahaahahaha 

In fact, maybe we shouldn't explain that at all. Just remove that keep it simple. What do you think? :-)

michelav replied:
I have removed it. ;-)

anayden commented on Lab1_Operationalizing_Pytorch_with_Mlflow/README.md:
```diff
@@ -1,31 +1,127 @@
+| 2 | Durable Data Storage | Every serious solution needs to store durable data | Tracking Server | Google Cloud SQL | A Postgres Database hosted by Google Cloud SQL is reliable and fast enough for our needs  |
+| 3 | Large and Binary Data Storage | ML projects eventually need to store models and extra large files, This type of data doesn't fit in RDBMS | Tracking Server<br>Artifact Store | Google Cloud Storage | Same as #1 |
+| 4 | Shipment and Deployment | There are many deployable assets as well as acessory tools that need specific environment and SO libraries to run. Just installing them into a machine isn't a viable option | Training  Node<br>Google Cloud Services | Docker | It's the _de facto_ pattern for shipping and deploying things |
+| 5 | Lib Management | Python has a powerful but sometimes confusing and conflicting library ecossystem. Our components may depend on conflicting libs and that will lead to problems in Production | All Nodes and Machines | Conda<br>Pip | Both are heavily used by Python community. They also work quite well with isolation provided by Docker images |
```
We don't use Conda anymore, do we?

michelav replied:
Oooops. Removed!

anayden commented on Lab1_Operationalizing_Pytorch_with_Mlflow/README.md:
```diff
@@ -1,31 +1,127 @@
+| 5 | Lib Management | Python has a powerful but sometimes confusing and conflicting library ecossystem. Our components may depend on conflicting libs and that will lead to problems in Production | All Nodes and Machines | Conda<br>Pip | Both are heavily used by Python community. They also work quite well with isolation provided by Docker images |
+| 6 | Remote Access | Sometimes we need to issue remote commands or operate a remote machine | All Nodes and Machines | SSH | Hey... It's SSH. There's no need to rationale :-P |
+| 7 | CI/CD & Workflow Management | We need to seamlessly integrate and deploy ML stuff. We also have to coordinate the process between nodes | All Items | Github Actions | We're using GitHub, it's heavily used by community and comes off-the-shelf |
+| 8 | Scheduler | Some activities are triggered by clock | All Nodes and Machines | CRON | We are using Unix like machines |
```
Just wondering what we use cron for

michelav replied:
Nothing. Forgot it there. We thought it would be of use, but we can do everything with GH + gcloud.

anayden commented on Lab1_Operationalizing_Pytorch_with_Mlflow/README.md:
```diff
@@ -1,31 +1,127 @@
+Actions (GH Actions);
+ 
+ 1. ML pipeline starts. we are considering that previous data activities (Extraction, Validation and
+ Preparation) were already executed. GH Actions issues a _gcloud commandIn to _Training Node_, so the 
```
minor: I think it should be `_gcloud_ command to _Training Node_`. Or something like that?

michelav replied:
Forgot that. Fixed!

anayden reviewed (approved):
Michael, I like your writeup sooooo much! A have a few super minor comments, but overall I'm very impressed. Well done!

michelav committed:
Fix typos

michelav committed:
Fix info about how to access models in torchserve

Change web access to access via HTTP since old info could lead someone to think a WEB UI would be available.
Although torchserve allows HTTPS access as well, it's not used in the project.

michelav committed:
Remove unused architectural mechanisms

There is no need to use Conda since Pip works quite well with Docker and completely satisfy our needs.
Prior, we chose Cron for scheduling because we thought there would be this kind of activity in conjunction with 
GitHub Runners. Since GCloud GitHub Actions address our problems, removing the scheduling mechanism.

michelav commented:
> Michael, I like your writeup sooooo much! A have a few super minor comments, but overall I'm very impressed. Well done!

feeling flattered. :grin: It got much better after your comments. Tks!

savagej commented on Lab1_Operationalizing_Pytorch_with_Mlflow/README.md:
```diff
@@ -1,31 +1,126 @@
+
+![Solution Workflow](imgs/workflow.svg)
+
+The first stage is Development Workflow. It concerns the continuous integration of new ML code and Dataset 
+changes into a Development Server. Here follows the steps:
```
Does it makes sense here or before here to talk about the experimentation work that can be done in MLflow before we get to the point of productionisation? I only mention it since the strong part of MLflow is that a (potentially cowboy!) data scientist can do all their experimentation work in MLflow, track parameters, compare the performance etc. Then they decide what to commit to master, and the MLops pipeline we've built out here takes over, but lets them see the progress in their "own" tool.
```suggestion
Before this MLops workflow is used, there would be an experimentation workflow used by a data scientist. They would be able to use this repository to test out various models, architectures, hyperparameters etc. which would all be logged and tracked in MLflow available for them to compare the performance in the UI. Once they had decided on a new model to move to production, they can create a Pull Request, ready for their team to review, which would then move into the MLops workflow proper.
The first stage is Development Workflow. It concerns the continuous integration of new ML code and Dataset 
changes into a Development Server. Here follows the steps:
```

michelav replied:
John... you're totally right. It's a great suggestion. I thought it would be nice if we could also mention this during the context and give MLFlow a nice spot. So, I did quite a big change in the text. Don't know if its too formal for our purposes. Anyway, take a look and if you think it doesn't fit there we may rollback it and accept your text right away. :wink:

savagej commented on Lab1_Operationalizing_Pytorch_with_Mlflow/README.md:
````diff
@@ -1,31 +1,126 @@
+ 
+ 1. The development workflow ends by deploying the trained model in Google Cloud Run Service. 
+ GH Actions creates a new image which contains the trained model (```mlflow create deployment```) which
+ is pushed to Google Container Registry.
````
```suggestion
 is pushed to Google Container Registry. This image is deployed to a development Cloud Run deployment, where various acceptance tests can be performed to ensure the image is ready for production.
```

savagej commented on Lab1_Operationalizing_Pytorch_with_Mlflow/README.md:
````diff
@@ -1,31 +1,126 @@
+ Storage Service acessible from Internet;
+ 
+ 1. The development workflow ends by deploying the trained model in Google Cloud Run Service. 
+ GH Actions creates a new image which contains the trained model (```mlflow create deployment```) which
````
```suggestion
 GH Actions creates a new image which contains the trained model (```mlflow create deployment```) which is tagged with the MLflow model name and version and then
```

savagej commented on Lab1_Operationalizing_Pytorch_with_Mlflow/README.md:
```diff
@@ -1,31 +1,126 @@
+
+The decision to promote a model to Production happens after some experiments. Model version _X_ is selected. This _external event_ triggers the Production Workflow:
+
+5. Operations and Product Management Team decides to deploy a model to Production environment. A release command is issued, triggering GitHub Actions. The runner pulls and configures the corresponding Docker image according to Production environment;
```
```suggestion
5. Operations and Product Management Team decides to deploy a model to Production environment. A release command is issued, specifying a model name and version to deploy, triggering GitHub Actions. The runner pulls and configures the corresponding Docker image and deploys this to the Production Cloud Run environment
```

savagej reviewed (approved):
Epic!! This is unbelievable, we're lucky to have this as the "face" of the project. This is the first thing people will see so it's great that it's so comprehensive and shiny too!

michelav committed:
Detail model deployment process regarding MLFlow tracking info

Co-authored-by: John Savage <dtjohnnyb@gmail.com>

michelav committed:
Describe MLFlow capabilities when  Experimenting

Our solution focuses on Production pipeline. We must let clear that MLFlow excels at Experimentation
and that we could address this in the future.

michelav committed:
Fix Typo

savagej commented on Lab1_Operationalizing_Pytorch_with_Mlflow/README.md:
```diff
@@ -1,31 +1,140 @@
+MLflow is an open source platform to manage the ML lifecycle, including experimentation, reproducibility, 
+deployment, and a central model registry. One may interact with its components by command line interface
+or well-known APIs (Python, R, Java and REST), building a workflow capable of handling development and
+production activies. Actually, Its simple design allows small teams (or a lone cowboy data scientist as well) 
```
Ha, cowboy might be a bit insulting out of context!!
```suggestion
production activies. Actually, Its simple design allows small teams (or a lone data scientist as well) 
```

michelav replied:
"Cowboy" removed. Still wanna be a cowboy something... maybe bc old western movies... haha

michelav committed:
Remove confuse explanation about Monitoring Node
